src/librustc_mir/borrow_check.rs

```diff
@@ -1115,13 +1116,238 @@ enum NoMovePathFound {
+                        Lvalue::Projection(box Projection { base, elem }) => (base, elem),
+                        _ => bug!("lvalue has no base?")
+                    };
+                    let base_ty = base.ty(self.mir, self.tcx).to_ty(self.tcx);
+
+                    match (elem, &base_ty.sty, access) {
+                        (_, _, Shallow(Some(ArtificialField::Discriminant))) |
+                        (_, _, Shallow(Some(ArtificialField::ArrayLength))) => {
+                            // The discriminant and array length are like
+                            // additional fields on the type; they do not
+                            // overlap any existing data there. Furthermore,
+                            // they cannot actually be a prefix of any
+                            // borrowed lvalue (at least in MIR as it is
+                            // currently.)
+                            debug!("borrow_conflicts_with_lvalue: implicit field");
+                            return false;
+                        }
+
+                        (ProjectionElem::Deref, _, Shallow(None)) => {
+                            debug!("borrow_conflicts_with_lvalue: shallow access behind ptr");
+                            return false;
+                        }
+                        (ProjectionElem::Deref, ty::TyRef(_, ty::TypeAndMut {
+                            ty: _, mutbl: hir::MutImmutable
+                        }), _) => {
+                            // the borrow goes through a dereference of a shared reference.
+                            //
+                            // I'm not sure why we are tracking these borrows - shared
+                            // references can *always* be aliased, which means the
+                            // permission check already account for this borrow.
+                            debug!("borrow_conflicts_with_lvalue: behind a shared ref");
+                            return false;
+                        }
+
+                        (ProjectionElem::Deref, _, Deep) |
+                        (ProjectionElem::Field { .. }, _, _) |
+                        (ProjectionElem::Index { ..}, _, _) |
+                        (ProjectionElem::ConstantIndex { .. }, _, _) |
+                        (ProjectionElem::Subslice { .. }, _, _) |
+                        (ProjectionElem::Downcast { .. }, _, _) => {
+                            // recursive case, let's see if we have a deref.
+                        }
+                    }
+                }
+                (Some(borrow_c), Some(lvalue_c)) => {
+                    match self.lvalue_element_conflict(borrow_c, lvalue_c) {
+                        Overlap::Arbitrary => {
+                            // the borrow partially overlaps. Give up - we
+                            // could try to be smarter here (e.g. counting
+                            // dereferences) but I think we should not.
```

Comment:
Comment: This primarily occurs with unions. For example, a borrow of `some_union.x.y` and `some_union.a` would trigger this case (when accessing the `x` and `a` fields).